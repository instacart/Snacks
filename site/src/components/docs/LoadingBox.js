import React, {Fragment} from 'react'
import Playground from '../Playground'

export default () => (

Light:

<Playground>
  {`
    <div>
      <LoadingBox background='light' />
    </div>
  `}
</Playground>

Standard:
<Playground>
  {`
    <div>
      <LoadingBox />
    </div>
    `}
  </Playground>

Dark:
<Playground>
  {`
    <div>
      <LoadingBox background='dark' />
    </div>
    `}
  </Playground>

Circle:
<Playground>
  {`
    <div>
      <LoadingBox shape='circle' background='dark' size={150} />
    </div>
    `}
  </Playground>

Square:
<Playground>
  {`
    <div>
      <LoadingBox shape='square' background='light' size={150} />
    </div>
    `}
  </Playground>

Line:
<Playground>
  {`
    <div>
      <LoadingBox shape='line' />
      <LoadingBox shape='line' />
      <LoadingBox shape='line' />
      <LoadingBox shape='line' />
    </div>
    `}
  </Playground>

Example Combination:
<Playground>
  {`

    const cardStyles = {
      display: 'flex',
      flexDirection: 'row',
      width: '280px'
    };
    const baseLineStyle = { marginTop: 10 };

    <div style={cardStyles}>
      <div style={{ flexGrow: 1 }}>
        <LoadingBox size={100} background='dark' />
        <LoadingBox shape='line' />
        <LoadingBox shape='line' />
        <LoadingBox shape='line' />
      </div>
      <div>
        <LoadingBox shape='circle' />
      </div>
    </div>
    `}
  </Playground>

**Note:** the size prop can be a percentage (supplied as a string) or a number in pixels:
<Playground>
  {`
    <div style={{ width: 200 }}>
      <LoadingBox shape='line' size='50%' />
      <LoadingBox shape='line' size={100} />
    </div>
  `}
</Playground>

)
