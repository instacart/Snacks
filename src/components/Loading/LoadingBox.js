import { keyframes } from '@emotion/core'
import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'
import { colors } from '../../styles'

const buildKeyFrames = (startColor, endColor) =>
  keyframes({
    '0%': { backgroundColor: startColor },
    '50%': { backgroundColor: endColor },
    '100%': { backgroundColor: startColor },
  })

const animations = {
  loading: buildKeyFrames(colors.GRAY_93, colors.GRAY_88),
  darkLoading: buildKeyFrames(colors.GRAY_88, colors.GRAY_74),
  lightLoading: buildKeyFrames(colors.GRAY_97, colors.GRAY_93),
}

const baseLoadingStyle = {
  animation: `${animations.loading} 3s ease-in-out infinite`,
  background: colors.GRAY_93,
  width: '100%',
  height: '20px',
}

const backgroundStyles = {
  dark: {
    animation: `${animations.darkLoading} 3s ease-in-out infinite`,
    background: colors.GRAY_88,
  },
  light: {
    animation: `${animations.lightLoading} 3s ease-in-out infinite`,
    background: colors.GRAY_97,
  },
}

const shapeStyles = {
  circle: {
    borderRadius: '50%',
    height: 100,
    width: 100,
  },
  square: {
    height: 100,
    width: 100,
  },
  line: {
    marginTop: 10,
    marginBottom: 10,
    width: 150,
  },
}

const determineStyle = (background, shape, size) => {
  const sizeStyles = !shape || shape === 'line' ? { width: size } : { width: size, height: size }

  return {
    ...(backgroundStyles[background] || {}),
    ...(shapeStyles[shape] || {}),
    ...(size ? sizeStyles : {}),
  }
}

class LoadingBox extends PureComponent {
  static propTypes = {
    className: PropTypes.string,

    /** Use for rendering dark backgrounds. */
    background: PropTypes.oneOf(['light', 'dark']),

    /** Use for rendering light backgrounds, overrides dark */
    shape: PropTypes.oneOf(['circle', 'square', 'line']),

    /**
     *  By default, `size` will determine the components width.
     *
     *  If the `shape` is prop `circle` or `square`, `size` will apply to height and width.
     */
    size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

    /**
     * Optional style overrides merged into emotion css
     *
     * @deprecated
     * This prop exists for backwards compatibility and will be
     * removed in a future version
     */
    style: PropTypes.object,
  }

  render() {
    const { className, background, shape, size, style } = this.props

    const boxStyle = {
      ...baseLoadingStyle,
      ...determineStyle(background, shape, size),
      ...style,
    }

    return <div className={className} css={boxStyle} />
  }
}

export default LoadingBox
