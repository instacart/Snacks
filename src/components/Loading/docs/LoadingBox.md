Light:
```jsx
import { LoadingBox } from 'ic-snacks'

; <div>
    <LoadingBox background="light" />
  </div>
```

Standard:
```jsx
import { LoadingBox } from 'ic-snacks'

; <div>
    <LoadingBox />
  </div>
```

Dark:
```jsx
import { LoadingBox } from 'ic-snacks'

; <div>
    <LoadingBox background="dark" />
  </div>
```

Circle:
```jsx
import { LoadingBox } from 'ic-snacks'

; <div>
    <LoadingBox shape="circle" background="dark" size={150} />
  </div>
```

Square:
```jsx
import { LoadingBox } from 'ic-snacks'

; <div>
    <LoadingBox shape="square" background="light" size={150} />
  </div>
```

Line:
```jsx
import { LoadingBox } from 'ic-snacks'

; <div>
    <LoadingBox shape="line" />
    <LoadingBox shape="line" />
    <LoadingBox shape="line" />
    <LoadingBox shape="line" />
  </div>
```

Example Combination:

```jsx
import { LoadingBox } from 'ic-snacks'

const cardStyles = {
  display: 'flex',
  flexDirection: 'row',
  width: '280px',
}
const baseLineStyle = { marginTop: 10 }

; <div style={cardStyles}>
    <div style={{ flexGrow: 1 }}>
      <LoadingBox size={100} background="dark" />
      <LoadingBox shape="line" />
      <LoadingBox shape="line" />
      <LoadingBox shape="line" />
    </div>
    <div>
      <LoadingBox shape="circle" />
    </div>
  </div>
````

**Note:** the size prop can be a percentage (supplied as a string) or a number in pixels:

```jsx
import { LoadingBox } from 'ic-snacks'

; <div style={{ width: 200 }}>
    <LoadingBox shape="line" size="50%" />
    <LoadingBox shape="line" size={100} />
  </div>
```
