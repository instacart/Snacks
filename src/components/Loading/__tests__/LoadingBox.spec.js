import React from 'react'
import renderer from 'react-test-renderer'
import LoadingBox from '../LoadingBox'

it('renders the standard LoadingBox correctly', () => {
  const tree = renderer.create(<LoadingBox style={{ width: '50px' }} />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders the line LoadingBox', () => {
  const tree = renderer.create(<LoadingBox shape="line" />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders the circle LoadingBox', () => {
  const tree = renderer.create(<LoadingBox shape="circle" />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders the square LoadingBox', () => {
  const tree = renderer.create(<LoadingBox shape="square" />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders the dark LoadingBox', () => {
  const tree = renderer.create(<LoadingBox background="dark" />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders the light LoadingBox', () => {
  const tree = renderer.create(<LoadingBox background="light" />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders the LoadingBox with a size', () => {
  const tree = renderer.create(<LoadingBox size={100} />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders the LoadingBox with a combination of props', () => {
  const tree = renderer.create(<LoadingBox size={200} background="dark" shape="circle" />).toJSON()
  expect(tree).toMatchSnapshot()
})
