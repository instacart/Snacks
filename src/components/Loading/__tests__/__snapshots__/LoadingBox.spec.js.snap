// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the LoadingBox with a combination of props 1`] = `
<div
  data-radium={true}
>
  <div
    data-radium={true}
    style={
      Object {
        "animation": "x 3s ease-in-out infinite",
        "animationName": "darkLoading-radium-animation-f5235588",
        "backgroundColor": "#E0E0E0",
        "borderRadius": "50%",
        "height": 200,
        "width": 200,
      }
    }
  />
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "@keyframes darkLoading-radium-animation-f5235588 {
0%{background-color: #E0E0E0;}
50%{background-color: #BDBDBD;}
100%{background-color: #E0E0E0;}
}
",
      }
    }
  />
</div>
`;

exports[`renders the LoadingBox with a size 1`] = `
<div
  data-radium={true}
>
  <div
    data-radium={true}
    style={
      Object {
        "animation": "x 3s ease-in-out infinite",
        "animationName": "loading-radium-animation-2dabe53b",
        "backgroundColor": "#EEEEEE",
        "height": "20px",
        "width": 100,
      }
    }
  />
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "@keyframes loading-radium-animation-2dabe53b {
0%{background-color: #EEEEEE;}
50%{background-color: #E0E0E0;}
100%{background-color: #EEEEEE;}
}
",
      }
    }
  />
</div>
`;

exports[`renders the circle LoadingBox 1`] = `
<div
  data-radium={true}
>
  <div
    data-radium={true}
    style={
      Object {
        "animation": "x 3s ease-in-out infinite",
        "animationName": "loading-radium-animation-2dabe53b",
        "backgroundColor": "#EEEEEE",
        "borderRadius": "50%",
        "height": 100,
        "width": 100,
      }
    }
  />
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "@keyframes loading-radium-animation-2dabe53b {
0%{background-color: #EEEEEE;}
50%{background-color: #E0E0E0;}
100%{background-color: #EEEEEE;}
}
",
      }
    }
  />
</div>
`;

exports[`renders the dark LoadingBox 1`] = `
<div
  data-radium={true}
>
  <div
    data-radium={true}
    style={
      Object {
        "animation": "x 3s ease-in-out infinite",
        "animationName": "darkLoading-radium-animation-f5235588",
        "backgroundColor": "#E0E0E0",
        "height": "20px",
        "width": "100%",
      }
    }
  />
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "@keyframes darkLoading-radium-animation-f5235588 {
0%{background-color: #E0E0E0;}
50%{background-color: #BDBDBD;}
100%{background-color: #E0E0E0;}
}
",
      }
    }
  />
</div>
`;

exports[`renders the light LoadingBox 1`] = `
<div
  data-radium={true}
>
  <div
    data-radium={true}
    style={
      Object {
        "animation": "x 3s ease-in-out infinite",
        "animationName": "lightLoading-radium-animation-438c49ee",
        "backgroundColor": "#F7F7F7",
        "height": "20px",
        "width": "100%",
      }
    }
  />
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "@keyframes lightLoading-radium-animation-438c49ee {
0%{background-color: #F7F7F7;}
50%{background-color: #EEEEEE;}
100%{background-color: #F7F7F7;}
}
",
      }
    }
  />
</div>
`;

exports[`renders the line LoadingBox 1`] = `
<div
  data-radium={true}
>
  <div
    data-radium={true}
    style={
      Object {
        "animation": "x 3s ease-in-out infinite",
        "animationName": "loading-radium-animation-2dabe53b",
        "backgroundColor": "#EEEEEE",
        "height": "20px",
        "marginBottom": 10,
        "marginTop": 10,
        "width": 150,
      }
    }
  />
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "@keyframes loading-radium-animation-2dabe53b {
0%{background-color: #EEEEEE;}
50%{background-color: #E0E0E0;}
100%{background-color: #EEEEEE;}
}
",
      }
    }
  />
</div>
`;

exports[`renders the square LoadingBox 1`] = `
<div
  data-radium={true}
>
  <div
    data-radium={true}
    style={
      Object {
        "animation": "x 3s ease-in-out infinite",
        "animationName": "loading-radium-animation-2dabe53b",
        "backgroundColor": "#EEEEEE",
        "height": 100,
        "width": 100,
      }
    }
  />
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "@keyframes loading-radium-animation-2dabe53b {
0%{background-color: #EEEEEE;}
50%{background-color: #E0E0E0;}
100%{background-color: #EEEEEE;}
}
",
      }
    }
  />
</div>
`;

exports[`renders the standard LoadingBox correctly 1`] = `
<div
  data-radium={true}
>
  <div
    data-radium={true}
    style={
      Object {
        "animation": "x 3s ease-in-out infinite",
        "animationName": "loading-radium-animation-2dabe53b",
        "backgroundColor": "#EEEEEE",
        "height": "20px",
        "width": "50px",
      }
    }
  />
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "@keyframes loading-radium-animation-2dabe53b {
0%{background-color: #EEEEEE;}
50%{background-color: #E0E0E0;}
100%{background-color: #EEEEEE;}
}
",
      }
    }
  />
</div>
`;
